R/brier_score.R

```diff
@@ -0,0 +1,241 @@
+    assert_numeric(events, any.missing = FALSE, finite = TRUE)
+    assert_that(
+        length(times) == length(events),
+        all(events == 1 | events == 0)
+    )
+    events_cen <- 1 - events
+    ord <- order(times, events_cen)
+    times <- times[ord]
+    events <- events[ord]
+    events_cen <- events_cen[ord]
+    cs_events <- cumsum(events)
+    cs_censor <- cumsum(events_cen)
+
+    g_times <- unique(times)
+    g_n_events_cen <- tapply(events_cen, times, sum)
+    g_cs_events_cen <- tapply(cs_censor, times, max)
+    g_cs_events <- tapply(cs_events, times, max)
+    g_is_cen <- tapply(events_cen, times, max)
+
+    nrisk <- length(times) - g_cs_events - g_cs_events_cen + g_n_events_cen
+    surv_interval <- 1 - g_n_events_cen / nrisk
+    surv_interval <- ifelse(nrisk == 0, 1, surv_interval)
+    surv <- cumprod(surv_interval)
+
+    ct <- g_times[which(g_is_cen == 1)]
+    sv <- surv[which(g_is_cen == 1)]
+
+    names(surv) <- NULL
+    names(ct) <- NULL
+    names(sv) <- NULL
+    list(
+        t = t,
+        surv = c(1, sv)[findInterval(t, ct) + 1]
+    )
+}
+
+
+#' @rdname reverse_km
+reverse_km_cen_first <- function(t, times, events) {
+    assert_numeric(t, any.missing = FALSE, finite = TRUE)
+    assert_numeric(times, any.missing = FALSE, finite = TRUE)
+    assert_numeric(events, any.missing = FALSE, finite = TRUE)
+    assert_that(
+        length(times) == length(events),
+        all(events == 1 | events == 0)
+    )
+    dat <- data.frame(
+        times = times,
+        events = events,
+        cen_events = 1 - events
+    )
+    mod <- survival::survfit(
+        survival::Surv(times, cen_events) ~ 1,
+        data = dat
+    )
+    preds <- summary(mod, times = t, extend = TRUE)$surv
+
+    assert_that(
+        length(preds) == length(t)
+    )
+    list(
+        t = t,
+        surv = preds[order(order(t))]
+    )
+}
+
+
+
+#' Brier Score
+#'
+#' @inheritParams Brier-Score-Shared
+#'
+#' @description
+#' Implements the Brier Score as detailed in Blanche et. al. (2015)
+#'
+#' @details
+#' - `bs_get_squared_dist()` - implements the squared distance part
+#' of the formula.
+#' - `bs_get_weights()` - implements the IPCW weighting
+#'
+#' @keywords internal
+brier_score <- function(
+    t,
+    times,
+    events,
+    pred_mat,
+    maintain_cen_order = TRUE,
+    event_offset = TRUE
+) {
+
+    square_diff_mat <- bs_get_squared_dist(
+        t = t,
+        times = times,
+        events = events,
+        pred_mat = pred_mat
+    )
+
+    weight_mat <- bs_get_weights(
+        t = t,
+        times = times,
+        events = events,
+        event_offset = event_offset,
+        maintain_cen_order = maintain_cen_order
+    )
+
+    # the following is a computational shortcut for diag(A %*% B)
+    # as we don't want to compute the off-diagonal entries of the
+    # matrix multiplication
+    x <- colSums(weight_mat * square_diff_mat)
+    names(x) <- t
+    return(x / length(times))
```

Comment:
```suggestion
    x / length(times)
```
good to avoid explicit return where not needed